Game crash while starting (BadWindow (invalid Window parameter))

Hi guys!

## System Information

- Operating System:  Lubuntu 18.04 (fresh install)
- OpenRA Version: `release-20200503`
- OpenRA Mod: Red Alert and Tiberian Dawn
- Source: Official AppImage

## Exception log

```
epoc@Laptop-Epoc:~/Documents$ chmod a+x OpenRA-Red-Alert-x86_64.AppImage
epoc@Laptop-Epoc:~/Documents$ ./OpenRA-Red-Alert-x86_64.AppImage
Platform is Linux
Engine version is release-20200503
X Error of failed request:  BadWindow (invalid Window parameter)
  Major opcode of failed request:  4 (X_DestroyWindow)
  Resource id in failed request:  0x360001d
  Serial number of failed request:  281
  Current serial number in output stream:  282
Red Alert has encountered a fatal error.
Please refer to the crash logs and FAQ for more information.

Log files are located in ~/.openra/Logs
The FAQ is available at http://wiki.openra.net/FAQ
```

Same error log for Tiberian Dawn.

## Additional information

The aforementioned logs location (`~/.openra/Logs`) doesn't exist. I found them in `~/.config/openra/Logs/`, however everything's empty:

```
epoc@Laptop-Epoc:~/Documents$ ll ~/.config/openra/Logs/
total 8
drwxrwxr-x 2 epoc epoc 4096 mai    8 21:43 ./
drwxrwxr-x 3 epoc epoc 4096 mai    8 21:43 ../
-rw-rw-r-- 1 epoc epoc    0 mai    9 11:09 debug.log
-rw-rw-r-- 1 epoc epoc    0 mai    9 11:09 geoip.log
-rw-rw-r-- 1 epoc epoc    0 mai    9 11:09 graphics.log
-rw-rw-r-- 1 epoc epoc    0 mai    9 11:09 nat.log
-rw-rw-r-- 1 epoc epoc    0 mai    9 11:09 perf.log
-rw-rw-r-- 1 epoc epoc    0 mai    9 11:09 server.log
-rw-rw-r-- 1 epoc epoc    0 mai    9 11:09 sound.log
```

Furthermore, simply switching the order of `supportedProfiles` to call `CanCreateGLWindow()` on `Legacy` before `Embedded` results in the following error:
```
Platform is Linux
Engine version is release-20200503
Using SDL 2 with OpenGL (Legacy) renderer
Desktop resolution: 1920x1080
No custom resolution provided, using desktop resolution
Using resolution: 1920x1080
Using window scale 1.00
Exception of type `System.InvalidOperationException`: Can not create OpenGL context. (Error: Invalid window)
  at OpenRA.Platforms.Default.Sdl2GraphicsContext.InitializeOpenGL () [0x0005f] in <004436c29d37490ba278b053e61e98ae>:0 
  at OpenRA.Platforms.Default.ThreadedGraphicsContext.RenderThread (System.Object contextObject) [0x0002a] in <004436c29d37490ba278b053e61e98ae>:0 
  at System.Threading.ThreadHelper.ThreadStart_Context (System.Object state) [0x00025] in <a17fa1457c5d44f2885ac746c1764ea5>:0 
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00071] in <a17fa1457c5d44f2885ac746c1764ea5>:0 
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00000] in <a17fa1457c5d44f2885ac746c1764ea5>:0 
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state) [0x0002b] in <a17fa1457c5d44f2885ac746c1764ea5>:0 
  at System.Threading.ThreadHelper.ThreadStart (System.Object obj) [0x0000f] in <a17fa1457c5d44f2885ac746c1764ea5>:0 

[ERROR] FATAL UNHANDLED EXCEPTION: System.InvalidOperationException: Can not create OpenGL context. (Error: Invalid window)
  at OpenRA.Platforms.Default.Sdl2GraphicsContext.InitializeOpenGL () [0x0005f] in <004436c29d37490ba278b053e61e98ae>:0 
  at OpenRA.Platforms.Default.ThreadedGraphicsContext.RenderThread (System.Object contextObject) [0x0002a] in <004436c29d37490ba278b053e61e98ae>:0 
  at System.Threading.ThreadHelper.ThreadStart_Context (System.Object state) [0x00025] in <a17fa1457c5d44f2885ac746c1764ea5>:0 
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00071] in <a17fa1457c5d44f2885ac746c1764ea5>:0 
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00000] in <a17fa1457c5d44f2885ac746c1764ea5>:0 
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state) [0x0002b] in <a17fa1457c5d44f2885ac746c1764ea5>:0 
  at System.Threading.ThreadHelper.ThreadStart (System.Object obj) [0x0000f] in <a17fa1457c5d44f2885ac746c1764ea5>:0 
```
I'll keep digging, but any suggestions are welcome.

--- a/AUTHORS
+++ b/AUTHORS
@@ -40,6 +40,7 @@ Also thanks to:
     * Biofreak
     * Braxton Williams (Buddytex)
     * Brendan Gluth (Mechanical_Man)
+    * Brent Gardner (bggardner)
     * Bryan Wilbur
     * Bugra Cuhadaroglu (BugraC)
     * Chris Cameron (Vesuvian)

--- a/OpenRA.Platforms.Default/OpenRA.Platforms.Default.csproj
+++ b/OpenRA.Platforms.Default/OpenRA.Platforms.Default.csproj
@@ -29,7 +29,7 @@
     <ProjectReference Include="..\OpenRA.Game\OpenRA.Game.csproj" />
     <PackageReference Include="OpenRA-Freetype6" Version="1.0.4" />
     <PackageReference Include="OpenRA-OpenAL-CS" Version="1.0.16" />
-    <PackageReference Include="OpenRA-SDL2-CS" Version="1.0.26" />
+    <PackageReference Include="OpenRA-SDL2-CS" Version="1.0.27" />
     <PackageReference Include="StyleCop.Analyzers" Version="1.1.118" />
     <PackageReference Include="Microsoft.NETFramework.ReferenceAssemblies" Version="1.0.0" PrivateAssets="All" />
     <AdditionalFiles Include="../stylecop.json" />

--- a/OpenRA.Platforms.Default/Sdl2PlatformWindow.cs
+++ b/OpenRA.Platforms.Default/Sdl2PlatformWindow.cs
@@ -145,8 +145,6 @@ namespace OpenRA.Platforms.Default
 				if (Platform.CurrentPlatform == PlatformType.Windows)
 					SetProcessDPIAware();
 
-				SDL.SDL_Init(SDL.SDL_INIT_NOPARACHUTE | SDL.SDL_INIT_VIDEO);
-
 				// Decide which OpenGL profile to use.
 				// We first need to query the available profiles on Windows/Linux.
 				// On macOS, known/consistent OpenGL support is provided by the OS.
@@ -161,6 +159,10 @@ namespace OpenRA.Platforms.Default
 					throw new InvalidOperationException("No supported OpenGL profiles were found.");
 
 				profile = supportedProfiles.Contains(requestProfile) ? requestProfile : supportedProfiles.First();
+
+				// Note: This must be called after the CanCreateGLWindow checks above,
+				// which needs to create and destroy its own SDL contexts as a workaround for specific buggy drivers
+				SDL.SDL_Init(SDL.SDL_INIT_VIDEO);
 				SetSDLAttributes(profile);
 
 				Console.WriteLine("Using SDL 2 with OpenGL ({0}) renderer", profile);
@@ -487,6 +489,9 @@ namespace OpenRA.Platforms.Default
 		static bool CanCreateGLWindow(GLProfile profile)
 		{
 			// Implementation inspired by TestIndividualGLVersion from Veldrid
+
+			// Need to create and destroy its own SDL contexts as a workaround for specific buggy drivers
+			SDL.SDL_Init(SDL.SDL_INIT_VIDEO);
 			SetSDLAttributes(profile);
 
 			var flags = SDL.SDL_WindowFlags.SDL_WINDOW_HIDDEN | SDL.SDL_WindowFlags.SDL_WINDOW_OPENGL;
@@ -494,6 +499,7 @@ namespace OpenRA.Platforms.Default
 			if (window == IntPtr.Zero || !string.IsNullOrEmpty(SDL.SDL_GetError()))
 			{
 				SDL.SDL_ClearError();
+				SDL.SDL_Quit();
 				return false;
 			}
 
@@ -502,11 +508,13 @@ namespace OpenRA.Platforms.Default
 			{
 				SDL.SDL_ClearError();
 				SDL.SDL_DestroyWindow(window);
+				SDL.SDL_Quit();
 				return false;
 			}
 
 			SDL.SDL_GL_DeleteContext(context);
 			SDL.SDL_DestroyWindow(window);
+			SDL.SDL_Quit();
 			return true;
 		}
 

